src/babel-serializer/network/pcap.hh

```diff
@@ -4,13 +4,142 @@
+/// for(auto p : reader.packets())
+///     my_process_package_data(p.data());
+/// NOTE: You can provide your own raw data by providing a cc::span<std::byte const> instead of a filepath
+/// NOTE: Currently only supports little endian and microsecond timestamps
+class reader
+{
+public:
+    /// header of the entire pcap file
+    struct header
+    {
+        uint32_t magic_number;  ///< magic number
+        uint16_t version_major; ///< major version number
+        uint16_t version_minor; ///< minor version number
+        int32_t thiszone;       ///< GMT to local correction
+        uint32_t sigfigs;       ///< accuracy of timestamps
+        uint32_t snap_length;   ///< max length of captured packets, in octets
+        uint32_t network;       ///< data link type
+    };
+    /// single packet data
+    struct packet
+    {
+        uint32_t timestamp_sec = 0;     ///< timestamp seconds
+        uint32_t timestamp_usec = 0;    ///< timestamp microseconds
+        uint32_t original_size = 0;     ///< original length of packet (NOT the stored one)
+        cc::span<std::byte const> data; ///< raw data of the packet
+    };
+
+public: // ctor
+    /// allocating constructor that will read the whole file into memory
+    reader(cc::string_view filepath, error_handler on_error = default_error_handler)
+    {
+        _on_error = on_error;
+        _file_data = file::read_all_bytes(filepath, on_error); // should this be lazy instead?
+        _data = _file_data.data();
+        _size = _file_data.size();
+    }
+    /// non-allocating constructor that references into some other data
+    reader(cc::span<std::byte const> data, error_handler on_error = default_error_handler)
+    {
+        _on_error = on_error;
+        _data = data.data();
+        _size = data.size();
+    }
+
+private:
+    struct packet_iterator
+    {
+        struct packet_header
+        {
+            uint32_t ts_sec;   /* timestamp seconds */
+            uint32_t ts_usec;  /* timestamp microseconds */
+            uint32_t incl_len; /* number of octets of packet saved in file */
+            uint32_t orig_len; /* actual length of packet */
+        };
+
+        packet_iterator(cc::span<std::byte const> data, error_handler on_error) : _data{data}, _on_error{on_error}
+        {
+            if (!_data.empty())
+                _current_packet = read_packet();
+        }
+
+        packet operator*() { return _current_packet; }
+
+        packet_iterator& operator++()
+        {
+            _pos += sizeof(packet_header) + _current_packet.data.size();
+            if (_pos < _data.size())
+                _current_packet = read_packet();
+            return *this;
+        }
+
+        packet_iterator operator++(int)
+        {
+            auto i = *this;
+            operator++();
+            return i;
+        }
+
+        bool operator!=(cc::sentinel) const { return _pos < _data.size(); }
+
+    private:
+        packet read_packet();
```

Comment:
I'd like it better if this sort of functionality was laid bare, and the iterator implemented on top of it, not the other way around.
That way this new reader and the existing function could also share code, for example.

Reply:
The old one is completely replaced though, so I see no code duplication (yet).
In general I agree with you though.
I'd still keep it this way for now as we do not intend to make it public API yet and we can refactor that once we have an actual code duplication.

Reply:
Then I'll leave it as it is for now.